### PR TITLE
Add gsm8k dataset and methods (IO, CoT, CoT-SC)

### PR DESCRIPTION
**Pull request by fjfricke**

### CODE REVIEW

**Comment by fjfricke** [comment]
currently, only IO is implemented

**Review by AjayDextrous — DISMISSED** [review]

**Comment by AjayDextrous** [comment]
make sure you do a squash merge btw

**Comment by fjfricke** [comment]
i will. I will wait until merge when it is done for all methods

File: examples/gsm8k/gsm8k.py
Comment (AjayDextrous) [inline examples/gsm8k/gsm8k.py:53]: Hey, isn't this basically COT and not IO, or am I mistaken?

Comment (fjfricke) [reply examples/gsm8k/gsm8k.py:53]: I think this would actually be ReAct. But you are right. This is not directly IO. CoT would be to separate it into multiple thoughts in a row. Will think about it and change it accordingly.

Comment (fjfricke) [reply examples/gsm8k/gsm8k.py:53]: You are right it is not react but a form of CoT with all thoughts in one output.

Comment (fjfricke) [reply examples/gsm8k/gsm8k.py:53]: makes a big difference btw (io vs cot)

Comment (AjayDextrous) [reply examples/gsm8k/gsm8k.py:53]: @fjfricke EDIT: nvmind i misread your comment

**Comment by fjfricke** [comment]
added CoT-SC (currently working not on Ollama)

File: examples/gsm8k/gsm8k.py
Comment (AjayDextrous) [inline examples/gsm8k/gsm8k.py:216]: I think it'd be nice if we have a uniform naming scheme (but this is something we can do in the next iteration, since its only a minor thing)

File: examples/gsm8k/gsm8k.py
Comment (AjayDextrous) [inline examples/gsm8k/gsm8k.py:250]: Again a next iteration thing, we could use the common utils code for the results code which is repeated in all the modules

File: examples/gsm8k/gsm8k.py
Comment (AjayDextrous) [inline examples/gsm8k/gsm8k.py]: newline at the end of file
```suggestion
    logging.info(f"Spent {spent} out of {budget} budget.")
    
```

File: examples/gsm8k/utils.py
Comment (AjayDextrous) [inline examples/gsm8k/utils.py]:
```suggestion
        return 1.0
        
```
newline here as well

**Review by AjayDextrous — APPROVED** [review]

File: examples/gsm8k/gsm8k.py
Comment (fjfricke) [reply examples/gsm8k/gsm8k.py:216]: yes i will keep that comment unresolved

File: examples/gsm8k/gsm8k.py
Comment (fjfricke) [reply examples/gsm8k/gsm8k.py:250]: absolutely. will keep that unresolved as well

**Pull request merged**
